internal/mock/sqladmin.go

```diff
@@ -0,0 +1,231 @@
+// Copyright 2021 Google LLC
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     https://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mock
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"net/http"
+	"net/http/httptest"
+	"sync"
+	"time"
+
+	sqladmin "google.golang.org/api/sqladmin/v1beta4"
+)
+
+// HTTPClient returns an *http.Client, URL, and cleanup function. The http.Client is
+// configured to connect to test SSL Server at the returned URL. This server will
+// respond to HTTP requests defined, or return a 5xx server error for unexpected ones.
+// The cleanup function will close the server, and return an error if any expected calls
+// weren't received.
+func HTTPClient(requests ...*Request) (*http.Client, string, func() error) {
+	// Create a TLS Server that responses to the requests defined
+	s := httptest.NewTLSServer(http.HandlerFunc(
+		func(resp http.ResponseWriter, req *http.Request) {
+			for _, r := range requests {
+				if r.matches(req) {
+					r.handle(resp, req)
+					return
+				}
+			}
+			// Unexpected requests should throw an error
+			resp.WriteHeader(http.StatusNotImplemented)
+			// TODO: follow error format better?
+			resp.Write([]byte(fmt.Sprintf("unexpected request sent to mock client: %v", req)))
+		},
+	))
+	// cleanup stops the test server and checks for uncalled requests
+	cleanup := func() error {
+		s.Close()
+		for i, e := range requests {
+			if e.reqCt > 0 {
+				return fmt.Errorf("%d calls left for specified call in pos %d: %v", e.reqCt, i, e)
+			}
+		}
+		return nil
+	}
+
+	return s.Client(), s.URL, cleanup
+
+}
+
+// Request represents a HTTP request for a test Server to mock responses for.
+//
+// Use NewRequest to initialize new Requests.
+type Request struct {
+	sync.Mutex
```

Comment:
What is this mutex protecting? Could we put it directly above the field(s) is protects?

Reply:
It's protecting the whole struct, which is why I embedded it vs just over certain fields.